src/puppetlabs/puppetdb/middleware.clj

```diff
@@ -299,15 +300,44 @@
           (app (assoc req :body-string payload))
           (http/error-response (str "Missing required parameter 'payload'")))))))
 
+(defn fail-when-payload-too-large
+  "Middlware that will return a 413 failure when the content-length of
+  a POST is too large (more than half the max heap). Acts as a noop
+  when that is not the case"
+  [app max-command-failure? max-command-size]
+  (fn [req]
+    (if (= :post (:request-method req))
+      (let [length-in-bytes (request/content-length req)]
+
+        (if length-in-bytes
+          (log/infof "Processing command with a content-length of %s bytes" length-in-bytes )
+          (log/warn (str "No content length found for POST. "
+                         "POST bodies that are too large could cause memory-related failures.")))
+
+        (if (and length-in-bytes
+                 max-command-failure?
+                 (> length-in-bytes max-command-size))
+          (do
+            (log/warnf "content-length of command is %sMB and is larger than the maximum allowed command size of %sMB"
```

Comment:
When I think about what's a warning vs. error, I think about if the system is doing what the user told it to do (reject catalogs that are too large) and if it's something that should trip an monitoring alarm (i.e. database is down, that's an error, needs user intervention).

In this case, seems like I wouldn't want to be paged if a catalog was too large. I also did tell PuppetDB to reject those... I guess the question at that point is if we make it a warning, will users not see it and be alarmed...